### ISSUE TITLE
On Linting

### ISSUE THREAD

**Issue by ostera** [tldr-pages/tldr#278]
As per the comments on #276, I hereby start the discussion on linting the markdown in the pages as part of a `make lint` task currently being discussed on #277.

Ceremonial displays aside, I think that having a lint task that will ensure the pages are in good shape is fundamental.

Now calling @rubenvereecken to the stand. We'd love to hear your thoughts on this.

**Comment by rubenvereecken** [comment]
I'll just start kick it off by picturing _my_ ideal case:
 - Syntax errors can impossibly be committed.
 - Linting is part of the git procedure and does not require any extra work on the contributor's part.
 - Linting can be done explicitly to check for mistakes.
 - The contributor setup for tldr is minimal.

I think the last one is the hardest one to fulfill, but I feel it's necessary to maintain the beautiful ease of contributing to tldr. Before the index people could just edit pages on GitHub and do a PR from there (never tested it though) without ever having to explicitly touch git. Now we took away that ease but that was fairly alright since most contributors seem to be working locally anyway. 

Current list of development dependencies:
 - Make
 - Ruby

Both are hard dependencies right now and I could imagine some people not having Ruby. No worries, they could install it. Let's try to keep this list of requirements to a minimum though, that's one of my main concerns.

Okay enough about concerns. The reason I'm enthousiastic about a `make lint` is because it can be part of a pre-commit hook that aborts the commit with a helpful message if anything goes wrong during (but not restricted to) the linting. This should rid ourselves of a lot of the PR's we're getting that shouldn't be necessary.

I'm open to any suggestions regarding pre-commit checks like linting, just loath to add hard dependencies like gems *cough* without sufficient justification.

**Comment by ostera** [comment]
I take the hit :stuck_out_tongue: there. But unless anyone here is willing to write a markdown linter in Bash, I think we're stuck with a 3rd party solution for it :sweat_smile: 

Most systems will have a Ruby version capable of running the scripts we currently have. If anything, some systems would require to install build-essentials to get make (or XCode's CLI Tools on OS X if I recall correctly).

In any case, whether it's a node module, a ruby gem, or a python egg, the setup usually involves one more command – which can as well be abstracted into the `make setup` rule, making it just as easy as it is right now.

:+1: on how the lint chain works, and the outputs.

**Comment by rubenvereecken** [comment]
I figured there was no way around 3rd party libraries, just needed someone to tell me I guess. A case can be made (for my peace of mind at least) for adding dependencies; they should at least guarantee that no more ill-formatted pages come our way.

I'd be more than happy to check out any PR's in this direction :+1:.

**Comment by waldyrious** [comment]
I really would like the web-based workflow to remain possible. Can't we setup something like Travis to check PRs and return an error message explaining what needs to be changed? i.e. add an entry to the index, fix syntax, etc. The commit itself isn't the problem, since unless the PR is merged, possible errors won't affect anyone. If there's something missing, the submitter can manually perform the changes that the git pre-commit hook / make script would do, and add those as another commit to the PR. Perhaps they'd eventually decide that doing it in command line is actually more convenient, but in the mean time we don't shut off the random contributor who just wanted to chip in their two cents to improve tldr.

And of course, we can explain the necessary steps to make a clean PR in CONTRIBUTING.md, so that the Travis failure can be avoided in the first place.

**Comment by ostera** [comment]
I don't mind Travis, but I'd certainly be more comfortable knowing all commits have been proofread and linted before ever turning into pull requests than having pull requests that require manual work to be cleaned up before merging.

Random contributors certainly know how to do a commit, I can't vouch for all of them to know how to deal with Travis.

**Comment by rubenvereecken** [comment]
Travis popped into my mind as well, I've used it on a previous bigger OS project and I think it could handle it, though I didn't do the setup myself. @ostera from my experience it would just say something along the lines of "Travis build failed" after someone submits a PR in the Github UI itself, so no know-how needed except that someone neads to point out "go check Travis logs to see what went wrong" (or read it out for them).

As to contributing guidelines @waldyrious, `CONTRIBUTING.md` already explains about format which is all we can do for those that use the web-based flow.

I'll look into Travis some time soon, it'd be an interesting alternative workflow and we can still keep manual linting for those of us inclined to the command line.

**Comment by waldyrious** [comment]
> than having pull requests that require manual work to be cleaned up before merging.
Well, that manual work would be entirely up to the contributors themselves, not the maintainers -- and it would be a choice, too (especially if they read the contributing instructions first): mind you, I am not suggesting to abandon the git hook workflow, just not forcing it on contributors.

>  I can't vouch for all of them to know how to deal with Travis.
Not sure I get what you mean here. The way I see it, sending a PR, getting an automatic (color-coded and all!) status report by Travis, clicking the "details" link and reading the (hopefully informative, if we do it right) error message in the log is certainly more straightforward (or at least less error-prone) than having to install Ruby and run make to set up the git pre-commit hook in the command line. The very fact that the former happens entirely through the browser would be enough to put inexperienced more at ease than doing things in the command line.

I am not saying it is *easier* or *better* to do things in the browser, just that we should offer contributors the choice to become comfortable with the more efficient tools on their own terms, rather than enforcing a workflow on them.

**Comment by waldyrious** [comment]
> CONTRIBUTING.md already explains about format which is all we can do for those that use the web-based flow.

I meant adding some pointers about the Travis failures and how to solve (or even better, prevent) them. Nothing much more elaborate than what we have now, sure.

**Comment by rubenvereecken** [comment]
Uh-oh, I remember why Travis wasn't an option. We _need_ the pre-commit git hook to rebuild the index, a feature which is already merged and will soon be used by clients for whatever ends they desire. What we _would_ need then would be a server post-push hook that rebuilds the index and commits it every time stuff's changed. And I don't think that's an option.

**Comment by waldyrious** [comment]
That's why I mentioned that Travis could warn contributors that they would need to update the index manually :)

**Comment by rubenvereecken** [comment]
Alright, that could work. People who are only editing wouldn't need to update the index as long as it really is only an index that contains no additional information. The index would need to be reformatted into something more human-readable for those who do add new pages, but that's quite okay.

What happens if we decided to add tags to the index though, based on page content? Admitted, this won't be happening very soon but we did keep agree on a format for the index that could easily be extended.

**Comment by waldyrious** [comment]
> What happens if we decided to add tags to the index though, based on page content?

Um, why would that be a problem? Are you perhaps thinking about adding the tags automatically? I am not sure that would work better than a manually curated list of keywords, unless I'm misunderstanding how many tags you guys had in mind (i.e. a full list of terms that aren't common words like "the", "an", etc indeed would work better automatically).

If otherwise the tags were meant to be added manually, I fail to see how the manual update of the index would pose a problem.

**Comment by rprieto** [comment]
I was away for a while, and so much happened! It looks great, there's some great momentum & lots of ideas here! 

A `make lint` task sounds great, and the "ideal case" looks good. I'm not fussed about dependencies personally, as long as we stick to one (let's avoid Node + Ruby + Python :smile:). I also agree with all the above conversation, catching errors before they make it to a PR would be great, hence the pre-commit hook :+1: However ideally it should work for web-only contributors too, hence Travis on PRs. I don't see why these 2 options are mutually exclusive? Travis is really a 2min job to setup, happy to help out with that.

About the tags, there's conversations in flight talking about either a curated list, or gathering keywords from the examples themselves. But that's probably another conversation, since like the index it's actually state-changing - whereas a `lint` task is not.

**Comment by rubenvereecken** [comment]
Welcome back @rpieto :). Travis will do the job nicely for now and we can set it up as soon as we have `make lint`. It just won't quite cut it for anything state-changing (as you put it), so people will have to edit the index manually for now. 

Next up: actual code to do this :P.  #277 was very promising for backticks but if we go with a Markdown library we could cut out that part of the PR - alas.

**Comment by rubenvereecken** [comment]
Sorry for the mistap there. I'm feeling inclined to check out the implementation myself tomorrow, I can to Travis at the same time.

If anyone else would really like to try their hands on this I'll give it up :).

**Comment by rubenvereecken** [comment]
Sorry for the mistap there. I'm feeling inclined to check out the implementation myself tomorrow, I can to Travis at the same time.

If anyone else would really like to try their hands on this I'll give it up :). 

EDIT: If ever we really want to go crazy on this, it's quite possible to have a Web hook to a moderated server that rebuilds the index and any other chores that need doing and people won't be able to do manually, then push a commit back to GitHub with those changes.

**Comment by waldyrious** [comment]
> I don't see why these 2 options are mutually exclusive?

exactly, they aren't :)

> If ever we really want to go crazy on this ...

A commit-bot? That would be great indeed :) but if we do that, we'd probably want to have it point (in the commit message) to instructions explaining to the PR author how to go about producing all the changes in a single commit, both for educational reasons and in the interest of keeping the repo's commit history clean.

**Comment by rprieto** [comment]
Probably a big piece of work, but I like the idea of the commit bot (eventually)! Could it be smart enough to run on `tldr-pages/master`, and only trigger for either *straight-to-master* commits or *PR-merge* commits (not each commit of the merged PR)?

Also :+1: on Markdown validation to cover more than backticks.

**Comment by rubenvereecken** [comment]
I'll try to get around to it sometime today probably.

@rprieto I think we would have it hook on push, since that's the unifying event to PR merges and direct commits to master. It'd rerun the builders, see if anything's changed and commit that. To address @waldyrious's concern it could simply squash the generated change commit into the previous commit (or at least, I hope the libraries can because it's an interactive process normally).

**Comment by rubenvereecken** [comment]
I went with [markdownlint](https://github.com/mivok/markdownlint), a Ruby gem, because it allows customizing the style (also, @ostera hinted at it).

After running the linter on all pages I examined the output and compiled for us here a summary of the warnings or errors we already suffer from. I'll leave the list here, annotated with my opinions, for others to discuss if they don't agree with me.

- MD009: Trailing spaces. Let's keep it.
- MD012: Multiple consecutive blank lines. Yup, keep it.
- MD013: Line length (80 chars). Nah, we don't need this.
- MD018: No space after hash on a header. Well it is ugly so let's keep it.
- MD022: Header should be surrounded by blank lines. Well, not too big a deal to me.
- MD030: Spaces after list markers. Well they _are_ prettier. 
- MD038: Spaces inside code span elements. I _never_ put spaces inside code span elements.

To summarize, I'd suppress rules 13, 22 and 38 and keep all the others on. In total there are 40 rules.

(Thanks tldr for helping me with `find pages -exec mdl {} \; | grep -E "MD[0-9]*" -o | sort | uniq`)

Okay so in the end I noticed it actually doesn't lint _that very well_ since it allows missing backticks. I'll investigate a bit...

**Comment by ostera** [comment]
:+1: to these rules and loving the momentum!

**Comment by waldyrious** [comment]
I mostly agree with @rubenvereecken's rule choices, but I think rule 22 wouldn't hurt, and would keep it by the same rationale as the other whitespace rules. But I don't feel too strongly about this, it's just a suggestion. :)

**Comment by rubenvereecken** [comment]
Cool, I'll get working then. There are nearly 250 infringements so it'll take some processing the current pages.

**Comment by ostera** [comment]
If we get the linting pushed, I think we can all chip in a bit with some pages?

**Comment by waldyrious** [comment]
I'm certainly willing to help :)

**Comment by rubenvereecken** [comment]
I'll close this one since I'm the one who kept this open. PR #309 finally rid us of all the remaining linting errors, courtesy of @lord63. Rejoice!
